### ISSUE TITLE
Reading and writing serialized Objects leads to errors

### ISSUE THREAD

**Issue by thebioengineer** [rstats-db/DBI#153]
I have been serializing objects and saving them into a db for later use. Since I updated to the newest DBI, I have not been able to deserialize the objects. I know my deserialization works because I am able to run it on the serialized object and it will deserialize properly. I have included a minimal reproducible example below 

```
library(DBI)
library(RSQLite)

SomeObjectToSave = "I should remain unchanged!"

###Here I serialize, compress, convert to a character vector and put the result into a data.frame in order to save into a db
serializeObject <- serialize(SomeObjectToSave, NULL)
compressedserializedData<-as.factor(rawToChar(memCompress(serializeObject,type = "bzip2"),TRUE))
serializeddata <- data.frame(serializedObject = compressedserializedData)

#### Put object into database ------------------------------
mydb <- dbConnect(RSQLite::SQLite(), dbname = "test.db")
dbWriteTable(mydb, "SerializedData", serializeddata, overwrite = FALSE)

###Now you should be able to immediatly read out that object and uncompress/serialize, right?
tmpCompressedObject <- dbReadTable(mydb, "SerializedData", row.names=FALSE)
dbDisconnect(mydb)

#### Convert Chars to Raws, making sure to put in the empty raws --------------
CompressedNonemptyRaws_position<-which(tmpCompressedObject$serializedObject!="")
CompressedNonEmptyRaws<-charToRaw(paste(tmpCompressedObject$serializedObject[CompressedNonemptyRaws_position],collapse=""))
CompressedrawVector <- raw(length(tmpCompressedObject$serializedObject))
CompressedrawVector[CompressedNonemptyRaws_position]<-CompressedNonEmptyRaws

#Decompress Raw vector
rawVector<-memDecompress(CompressedrawVector, type =  "bzip2")

#### Unserialize the object and return it -------------------------------------
unserializedObject<-unserialize(rawVector)

##### alas, this does not work, and I am not sure why. It has something to do with the reading the object from the db, because as you can see below, the deserialization works on the object when it is not saved to the db
CompressedNonemptyRaws_position2<-which(serializeddata$serializedObject!="")
CompressedNonEmptyRaws2<-charToRaw(paste(serializeddata$serializedObject[CompressedNonemptyRaws_position2],collapse=""))
CompressedrawVector2 <- raw(length(serializeddata$serializedObject))
CompressedrawVector2[CompressedNonemptyRaws_position2]<-CompressedNonEmptyRaws2

#Decompress Raw vector
rawVector2<-memDecompress(CompressedrawVector2, type =  "bzip2")

##### Unserialize the object and return it -------------------------------------
unserializedObject2<-unserialize(rawVector2)

testthat::expect_identical(unserializedObject2,SomeObjectToSave)
```

My sessionInfo() is below: 

```
sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] RSQLite_1.1-2 DBI_0.5-1    

loaded via a namespace (and not attached):
 [1] magrittr_1.5   R6_2.2.0       tools_3.3.2    crayon_1.3.2   memoise_1.0.0  Rcpp_0.12.8    digest_0.6.11  xlsxjars_0.6.1 testthat_1.0.2 rJava_0.9-8    xlsx_0.5.7    
[12] XML_3.98-1.5  
```

**Comment by krlmlr** [comment]
Thanks. I can confirm that this code worked with RSQLite 1.0.0. I need some time to understand why it worked before and doesn't work anymore. I assume this is a character encoding issue.

What is the context of the serialization? How important is it to keep the format stable? Would you consider rewriting the serialization to something that doesn't use a row (!) per table, but writes the data to blobs (using one row per object)?

**Comment by thebioengineer** [comment]
My specific use case for this is to store results from lengthy calculations in my shiny app. It stores the conditions in the name of the object so that if someone calls up the same conditions it can simply read that it already exists in the table and read the results out. 

I have some other code that saves all objects into a single table with two columns, the object name and the serialized data stored as a string. It followed a similar flow to above, but after converting the raw vector to a char, it creates a string by collapsing the characters and sending a write into table query with the data. I can create another min. Reproducible example with that structure, but i come across the same problem.

I do not know what a blob is, but I will look into it.

Thanks!

**Comment by krlmlr** [comment]
You create a blob column by passing a data frame with a column that is a list of raw objects: `data.frame(id = 1, data = I(list(serialize(..., NULL))))`. Alternatively, use the [blob](https://cran.r-project.org/web/packages/blob/index.html) class: `tibble::tibble(id = 1, data = blob::as.blob(serialize(..., NULL)))`. The documentation is weak here, but I'm working on improving it.

**Comment by thebioengineer** [comment]
Thanks for posting that! I was able to get that to work for part of my code. I have another part that uses dbGetQuery to build up a table of serialized objects, I did not originally include it because I thought that the core of the problem is related to the encoding, so the example i gave illustrated the problem

I think i stumbled upon the issue in the code. I am not sure if it has to do with DBI, or the fact that it stopped working was a coincidence. if i change the decompression sequence to include a part where the characters are encoded to the native locale, it works again:

`CompressedNonemptyRaws_position<-which(tmpCompressedObject$serializedObject!="")
CompressedNonEmptyRaws<-charToRaw(enc2native(paste(tmpCompressedObject$serializedObject[CompressedNonemptyRaws_position],collapse="")))
CompressedrawVector <- raw(length(tmpCompressedObject$serializedObject))
CompressedrawVector[CompressedNonemptyRaws_position]<-CompressedNonEmptyRaws

rawVector<-memDecompress(CompressedrawVector, type =  "bzip2")

unserializedObject<-unserialize(rawVector)
`


also, after closer inspection of rawToChar, there is this description:"...it is possible to create a result which is invalid in a multibyte locale, e.g. one using UTF-8." Is it possible that some of the characters when being read out were being converted to UTF-8?
